File: backend/app/routers/geo.py
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.auth import get_current_user, require_admin
from app.models.geo import Zone, District, Area, Cluster, ExamCenter
from app.schemas.geo import (
    ZoneCreate, ZoneUpdate, ZoneOut,
    DistrictCreate, DistrictUpdate, DistrictOut,
    AreaCreate, AreaUpdate, AreaOut,
    ClusterCreate, ClusterUpdate, ClusterOut,
    ExamCenterCreate, ExamCenterUpdate, ExamCenterOut,
)

router = APIRouter(prefix="/geo", tags=["Geo"])


# ── Zones ─────────────────────────────────────────────────────────────────────

@router.get("/zones", response_model=list[ZoneOut])
async def list_zones(db: AsyncSession = Depends(get_db), _=Depends(get_current_user)):
    result = await db.execute(select(Zone).order_by(Zone.name))
    return result.scalars().all()

@router.post("/zones", response_model=ZoneOut, status_code=201)
async def create_zone(body: ZoneCreate, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = Zone(**body.model_dump())
    db.add(obj)
    await db.commit()
    await db.refresh(obj)
    return obj

@router.get("/zones/{zone_id}", response_model=ZoneOut)
async def get_zone(zone_id: int, db: AsyncSession = Depends(get_db), _=Depends(get_current_user)):
    obj = await db.get(Zone, zone_id)
    if not obj:
        raise HTTPException(404, "Zone not found")
    return obj

@router.put("/zones/{zone_id}", response_model=ZoneOut)
async def update_zone(zone_id: int, body: ZoneUpdate, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = await db.get(Zone, zone_id)
    if not obj:
        raise HTTPException(404, "Zone not found")
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(obj, k, v)
    await db.commit()
    await db.refresh(obj)
    return obj

@router.delete("/zones/{zone_id}", status_code=204)
async def delete_zone(zone_id: int, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = await db.get(Zone, zone_id)
    if not obj:
        raise HTTPException(404, "Zone not found")
    await db.delete(obj)
    await db.commit()


# ── Districts ─────────────────────────────────────────────────────────────────

@router.get("/districts", response_model=list[DistrictOut])
async def list_districts(
    zone_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_user),
):
    q = select(District).order_by(District.name)
    if zone_id:
        q = q.where(District.zone_id == zone_id)
    result = await db.execute(q)
    return result.scalars().all()

@router.post("/districts", response_model=DistrictOut, status_code=201)
async def create_district(body: DistrictCreate, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = District(**body.model_dump())
    db.add(obj)
    await db.commit()
    await db.refresh(obj)
    return obj

@router.get("/districts/{district_id}", response_model=DistrictOut)
async def get_district(district_id: int, db: AsyncSession = Depends(get_db), _=Depends(get_current_user)):
    obj = await db.get(District, district_id)
    if not obj:
        raise HTTPException(404, "District not found")
    return obj

@router.put("/districts/{district_id}", response_model=DistrictOut)
async def update_district(district_id: int, body: DistrictUpdate, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = await db.get(District, district_id)
    if not obj:
        raise HTTPException(404, "District not found")
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(obj, k, v)
    await db.commit()
    await db.refresh(obj)
    return obj

@router.delete("/districts/{district_id}", status_code=204)
async def delete_district(district_id: int, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = await db.get(District, district_id)
    if not obj:
        raise HTTPException(404, "District not found")
    await db.delete(obj)
    await db.commit()


# ── Areas ─────────────────────────────────────────────────────────────────────

@router.get("/areas", response_model=list[AreaOut])
async def list_areas(
    district_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_user),
):
    q = select(Area).order_by(Area.name)
    if district_id:
        q = q.where(Area.district_id == district_id)
    result = await db.execute(q)
    return result.scalars().all()

@router.post("/areas", response_model=AreaOut, status_code=201)
async def create_area(body: AreaCreate, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = Area(**body.model_dump())
    db.add(obj)
    await db.commit()
    await db.refresh(obj)
    return obj

@router.get("/areas/{area_id}", response_model=AreaOut)
async def get_area(area_id: int, db: AsyncSession = Depends(get_db), _=Depends(get_current_user)):
    obj = await db.get(Area, area_id)
    if not obj:
        raise HTTPException(404, "Area not found")
    return obj

@router.put("/areas/{area_id}", response_model=AreaOut)
async def update_area(area_id: int, body: AreaUpdate, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = await db.get(Area, area_id)
    if not obj:
        raise HTTPException(404, "Area not found")
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(obj, k, v)
    await db.commit()
    await db.refresh(obj)
    return obj

@router.delete("/areas/{area_id}", status_code=204)
async def delete_area(area_id: int, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = await db.get(Area, area_id)
    if not obj:
        raise HTTPException(404, "Area not found")
    await db.delete(obj)
    await db.commit()


# ── Clusters ──────────────────────────────────────────────────────────────────

@router.get("/clusters", response_model=list[ClusterOut])
async def list_clusters(
    area_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_user),
):
    q = select(Cluster).order_by(Cluster.name)
    if area_id:
        q = q.where(Cluster.area_id == area_id)
    result = await db.execute(q)
    return result.scalars().all()

@router.post("/clusters", response_model=ClusterOut, status_code=201)
async def create_cluster(body: ClusterCreate, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = Cluster(**body.model_dump())
    db.add(obj)
    await db.commit()
    await db.refresh(obj)
    return obj

@router.get("/clusters/{cluster_id}", response_model=ClusterOut)
async def get_cluster(cluster_id: int, db: AsyncSession = Depends(get_db), _=Depends(get_current_user)):
    obj = await db.get(Cluster, cluster_id)
    if not obj:
        raise HTTPException(404, "Cluster not found")
    return obj

@router.put("/clusters/{cluster_id}", response_model=ClusterOut)
async def update_cluster(cluster_id: int, body: ClusterUpdate, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = await db.get(Cluster, cluster_id)
    if not obj:
        raise HTTPException(404, "Cluster not found")
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(obj, k, v)
    await db.commit()
    await db.refresh(obj)
    return obj

@router.delete("/clusters/{cluster_id}", status_code=204)
async def delete_cluster(cluster_id: int, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = await db.get(Cluster, cluster_id)
    if not obj:
        raise HTTPException(404, "Cluster not found")
    await db.delete(obj)
    await db.commit()


# ── ExamCenters ───────────────────────────────────────────────────────────────

@router.get("/exam-centers", response_model=list[ExamCenterOut])
async def list_exam_centers(
    district_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_user),
):
    q = select(ExamCenter).order_by(ExamCenter.name)
    if district_id:
        q = q.where(ExamCenter.district_id == district_id)
    result = await db.execute(q)
    return result.scalars().all()

@router.post("/exam-centers", response_model=ExamCenterOut, status_code=201)
async def create_exam_center(body: ExamCenterCreate, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = ExamCenter(**body.model_dump())
    db.add(obj)
    await db.commit()
    await db.refresh(obj)
    return obj

@router.get("/exam-centers/{ec_id}", response_model=ExamCenterOut)
async def get_exam_center(ec_id: int, db: AsyncSession = Depends(get_db), _=Depends(get_current_user)):
    obj = await db.get(ExamCenter, ec_id)
    if not obj:
        raise HTTPException(404, "ExamCenter not found")
    return obj

@router.put("/exam-centers/{ec_id}", response_model=ExamCenterOut)
async def update_exam_center(ec_id: int, body: ExamCenterUpdate, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = await db.get(ExamCenter, ec_id)
    if not obj:
        raise HTTPException(404, "ExamCenter not found")
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(obj, k, v)
    await db.commit()
    await db.refresh(obj)
    return obj

@router.delete("/exam-centers/{ec_id}", status_code=204)
async def delete_exam_center(ec_id: int, db: AsyncSession = Depends(get_db), _=Depends(require_admin)):
    obj = await db.get(ExamCenter, ec_id)
    if not obj:
        raise HTTPException(404, "ExamCenter not found")
    await db.delete(obj)
    await db.commit()

File: frontend/src/api/geo.ts
import api from "./client";

export interface Zone     { id: number; name: string; }
export interface District { id: number; name: string; zone_id: number; }
export interface Area     { id: number; name: string; district_id: number; }
export interface Cluster  { id: number; name: string; area_id: number; }

export const listZones = async (): Promise<Zone[]> => (await api.get("/geo/zones")).data;
export const listDistricts = async (zone_id?: number): Promise<District[]> => (await api.get("/geo/districts", { params: zone_id ? { zone_id } : {} })).data;
export const listAreas = async (district_id?: number): Promise<Area[]> => (await api.get("/geo/areas", { params: district_id ? { district_id } : {} })).data;
export const listClusters = async (area_id?: number): Promise<Cluster[]> => (await api.get("/geo/clusters", { params: area_id ? { area_id } : {} })).data;

export const createZone    = async (data: { name: string }): Promise<Zone>     => (await api.post("/geo/zones", data)).data;
export const updateZone    = async (id: number, data: Partial<Zone>): Promise<Zone>     => (await api.put(`/geo/zones/${id}`, data)).data;
export const deleteZone    = async (id: number): Promise<void>                           => { await api.delete(`/geo/zones/${id}`); };
export const createDistrict = async (data: { name: string; zone_id: number }): Promise<District> => (await api.post("/geo/districts", data)).data;
export const updateDistrict = async (id: number, data: Partial<District>): Promise<District> => (await api.put(`/geo/districts/${id}`, data)).data;
export const deleteDistrict = async (id: number): Promise<void>                               => { await api.delete(`/geo/districts/${id}`); };
export const createArea    = async (data: { name: string; district_id: number }): Promise<Area>  => (await api.post("/geo/areas", data)).data;
export const updateArea    = async (id: number, data: Partial<Area>): Promise<Area>    => (await api.put(`/geo/areas/${id}`, data)).data;
export const deleteArea    = async (id: number): Promise<void>                          => { await api.delete(`/geo/areas/${id}`); };
export const createCluster = async (data: { name: string; area_id: number }): Promise<Cluster>  => (await api.post("/geo/clusters", data)).data;
export const updateCluster = async (id: number, data: Partial<Cluster>): Promise<Cluster> => (await api.put(`/geo/clusters/${id}`, data)).data;
export const deleteCluster = async (id: number): Promise<void>                              => { await api.delete(`/geo/clusters/${id}`); };

// ── Exam Centers ─────────────────────────────────────────────────────────────
export interface ExamCenter { id: number; name: string; district_id?: number | null; street?: string | null; city?: string | null; state?: string | null; pincode?: string | null; }
export const listExamCenters = async (): Promise<ExamCenter[]> => (await api.get("/geo/exam-centers")).data;

// ── Lookups ──────────────────────────────────────────────────────────────────
export interface ExamCategory   { id: number; name: string; }
export interface NoExamReason   { id: number; reason: string; }
export interface NoAdmitReason  { id: number; reason: string; }

export const listExamCategories  = async (): Promise<ExamCategory[]>  => (await api.get("/lookups/exam-categories")).data;
export const listNoExamReasons   = async (): Promise<NoExamReason[]>  => (await api.get("/lookups/no-exam-reasons")).data;
export const listNoAdmitReasons  = async (): Promise<NoAdmitReason[]> => (await api.get("/lookups/no-admit-reasons")).data;

export interface Category    { id: number; name: string; }
export interface SubCategory { id: number; name: string; category_id: number; }

export const listCategories    = async (): Promise<Category[]>    => (await api.get("/lookups/categories")).data;
export const listSubCategories = async (category_id?: number): Promise<SubCategory[]> => {
  const params = category_id ? { category_id } : {};
  return (await api.get("/lookups/sub-categories", { params })).data;
};
